_posts/2022-06-08-#22778.md

```diff
@@ -0,0 +1,109 @@
+---
+layout: pr
+date: 2022-06-08
+title: "net processing: Reduce resource usage for inbound block-relay-only connections"
+pr: 22778
+authors: [jnewbery]
+components: ["p2p"]
+host: larryruane
+status: upcoming
+commit: 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4
+---
+
+## Notes
+
+- (These notes describe behavior with or without this PR.)
+
+- We are a `bitcoind` node. Each of our peers is either inbound (our
+  peer initiated the connection) or outbound (we initiated the connection).
+  Once the connection is established, the relationship is symmetric (the
+  protocol works the same in both directions).
+
+- When we initiate an outbound connection, we may decide that we won't
+  relay (forward) transactions to this peer.
+  This peer is designated a `block-relay-only` peer. The peer
+  doesn't know that we're treating it as `block-relay-only`.
+
+- When connecting with a peer, we allocate a `Peer` object.
+  If we'll need to notify and send transactions to
+  this peer (for example, if it's not a `block-relay-only` peer),
+  we create an additional `TxRelay` object to keep track
+  of state needed for relaying transactions to this peer.
+
+- The `TxRelay` object contains (in simplified form):
+   - the list of transactions (txids) that we know about
+     but haven't yet notified this peer about
+   - the "list" (actually a bloom filter) of transactions (txids) that
+     the peer knows about (because either we sent the transaction
+     to the peer, or it has sent it to us)
+   - if we've enabled BIP37, the bloom filter that the peer sends us
+     so that we can filter out transactions that it doesn't care about, 
+  
+- If we advertise support for BIP37 (bloom filter) to our peer
+  during version handshake, the peer can send us a bloom filter
+  after which we won't send transactions that aren't present in
+  the filter (to reduce bandwidth).
+
+- If a node connects to us (it's one of our inbound connections), it
+  can indicate that it doesn't want to receive transactions
+  from us. It does this using the `relay` field in the
+  [version](https://en.bitcoin.it/wiki/Protocol_documentation#version)
+  message.
+
+- If we've advertised support for bloom filters to this peer,
+  it may, at any time later, send us a `filterload` message,
+  which has the side-effect of enabling transaction relay from
+  us to this peer, if it had sent `relay = false` in the `version`
+  message.
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+1. What's a bloom filter?
+
+1. Why are some peers `block-relay-only`? Briefly, what's the purpose of having
+   some peers being `block-relay-only`?
+
+1. This PR reduces resource usage. Which resource, primarily? Why might this
+   be beneficial?
+
+1. Why is the `TxRelay` object separate from the `Peer` object?
```

Comment:
Please link to the declarations of these data structures. It helps specify exactly what you're referring to and gives attendees a hint for where to look in the codebase.